### PR TITLE
Introduce a parser error when the *same* custom op trait is used more than once

### PR DESCRIPTION
**Pull request by andreoss**
Regarding https://github.com/masak/007/issues/95

### CODE REVIEW

File: lib/_007/Parser/Actions.pm [hunk 1]
```diff
@@ -205,16 +209,12 @@ class _007::Parser::Actions {
             $op.ident.name.value.subst(/^ \w+ ":<"/, "").subst(/">" $/, "");
         }
 
-        sub tighter($op1, $op2) {
-            my $b = $*parser.oplevel.infixprec.first(*.contains(name($op1)), :k)
-                 > $*parser.oplevel.infixprec.first(*.contains(name($op2)), :k);
-            return $b;  # XXX: inexplicable runtime error if we return the value directly
+        sub tighter($op1, $op2, $_ = $*parser.oplevel.infixprec) {
+            .first(*.contains(name($op1)), :k) > .first(*.contains(name($op2)), :k);
```
Comment (vendethiel) [inline lib/_007/Parser/Actions.pm]: Maybe use `given` Instead?

File: lib/_007/Parser/Actions.pm [hunk 2]
```diff
@@ -168,7 +168,11 @@ class _007::Parser::Actions {
     }
 
     method traitlist($/) {
-        make Q::TraitList.new(:traits(Val::Array.new(:elements($<trait>».ast))));
+        my @traits = $<trait>».ast;
+        for bag( @traits.map: *.ident.name.value ).grep( *.value > 1 ) -> $p {
+             die X::Trait::Duplicate.new: :t($p.key)
```
Comment (vendethiel) [inline lib/_007/Parser/Actions.pm]: Not sure why the for, considering the `die` will prevent us from going forward?

File: lib/_007/Parser/Actions.pm (hunk 1 above)
Comment (andreoss) [inline lib/_007/Parser/Actions.pm]: Not even mine. Rebased it out.

**Comment by masak** [comment]
Thanks!

Will merge, but with two objections, one minor and one major:

* **Minor**: `$.t` is not such a descriptive name for an attribute. Should probably be `$.trait`. (I can see that the naming was probably based on `X::Trait::Conflict` which has `$.t1` and `$.t2`. Though it feels a bit more OK in that case, these should probably be renamed also.)

* **Major**: The patch goes a bit too far and forbids *all* duplicate traits. I see nothing inherently wrong with duplicating traits.

Let's say there's an `is wrapped(fn)` trait. Then this should be fine:

    sub foo() is wrapped(fn1) is wrapped(fn2) {
        # ...
    }

This is an architectural concern and not a merge-blocker; there's currently no way to declare traits in user-land, or to give custom traits any behavior, or to use traits for anything other than tweaking the properties of operators. I'm going to leave this one in as correct-for-the-moment and let future tests drive correct behavior.

**Pull request merged**